libs/MotionKit/source/MotionKit.cpp

```diff
@@ -58,61 +54,74 @@ void MotionKit::startStabilisation()
+		iterateOverRotation(euler_angles);
+	};
+
+	_imukit.onEulerAnglesReady(on_euler_angles_rdy_callback);
 }
 
 // LCOV_EXCL_START - Dynamic behavior, involving motors and time.
-void MotionKit::run()
+void MotionKit::iterateOverRotation(const interface::EulerAngles &angles, Rotation direction)
 {
-	auto last_yaw			= kReferenceAngle;
-	auto rotations_executed = 0;
+	auto must_stop = [&] { return !_stabilisation_requested && !_rotate_x_turns_requested && !_target_not_reached; };
 
-	auto must_rotate = [&] { return _rotate_x_turns_requested && rotations_executed != _rotations_to_execute; };
+	if (must_stop()) {
+		stop();
+		_imukit.onEulerAnglesReady({});
 
-	auto check_complete_rotations_executed = [&](auto current_yaw) {
-		if (std::abs(last_yaw - current_yaw) >= 300.F) {
-			++rotations_executed;
+		if (_on_rotation_ended_callback) {
+			_on_rotation_ended_callback();
 		}
-	};
 
-	while (must_rotate()) {
-		auto [current_pitch, current_roll, current_yaw] = _imukit.getEulerAngles();
-
-		check_complete_rotations_executed(current_yaw);
-
-		rtos::ThisThread::sleep_for(70ms);
-		last_yaw = current_yaw;
+		return;
 	}
 
-	_rotate_x_turns_requested = false;
-	_rotations_to_execute	  = 0;
+	auto counting_rotations = [&](auto current_yaw) {
+		if (auto abs_yaw_delta = std::abs(_euler_angles_previous.yaw - current_yaw); abs_yaw_delta >= 300.F) {
+			_angle_rotation_sum += 360.F - abs_yaw_delta;
+		} else {
+			_angle_rotation_sum += abs_yaw_delta;
+		}
+		_euler_angles_previous.yaw = current_yaw;
+	};
 
-	while (_stabilisation_requested || _target_not_reached) {
-		auto [pitch, roll, yaw] = _imukit.getEulerAngles();
-		auto [speed, rotation]	= _pid.processPID(pitch, roll, yaw);
+	counting_rotations(angles.yaw);
 
-		executeSpeed(speed, rotation);
+	if (_stabilisation_requested && _target_not_reached) {
+		auto [speed, rotation] = _pid.processPID(angles.pitch, angles.roll, angles.yaw);
 
-		rtos::ThisThread::sleep_for(70ms);
+		executeSpeed(speed, rotation);
 	}
 
-	if (_on_rotation_ended_callback != nullptr) {
-		_on_rotation_ended_callback();
+	// Without regulation
+	// if (_rotate_x_turns_requested && _target_not_reached) {
+	// 	if (auto error_position_current = _angle_rotation_target - _angle_rotation_sum;
+	// 		error_position_current > minimum_viable_position_error) {
+	// 		return;
+	// 	}
+	// 	_target_not_reached		  = false;
+	// 	_rotate_x_turns_requested = false;
+	// }
+
+	// With regulation
+	if (_rotate_x_turns_requested && _target_not_reached) {
+		auto speed = _pid.processPIDByError(_angle_rotation_target - _angle_rotation_sum);
+
+		executeSpeed(speed, direction);
 	}
-
-	_imukit.stop();
 }
 
 auto MotionKit::mapSpeed(float speed) const -> float
 {
-	return utils::math::map(speed, 0.F, kPIDMaxValue, kMinimalViableRobotPwm, kPwmMaxValue);
+	return utils::math::map(speed, 0.F, kEntryLimitSpeed, kMinimalViableRobotPwm, kPwmMaxValue);
 }
 
 void MotionKit::executeSpeed(float speed, Rotation direction)
```

Comment:
`executeSpeed` does not sound like real English

```suggestion
void MotionKit::setMotorsSpeedAndDirection(float speed, Rotation direction)
```